miniapps/tools/least-squares.cpp

```diff
@@ -0,0 +1,1720 @@
+         mass.AssembleElementMatrix2(fe_src_neighbor, fe_dst_neighbor, neighbor_trans,
+                                     neighbor_to_neighbor_mat);
+         src_neighbors_avg(i) = neighbor_to_neighbor_mat.InnerProduct(src_neighbor,
+                                                                      punity_dst_neighbor);
+
+         // e-DOFs to neighbor-DOFs (on neighbor) mass matrix
+         mass.AsymmetricElementMatrix(fe_dst_neighbor, fe_src_e, neighbor_trans, e_trans,
+                                      e_to_neighbor_mat, newton);
+
+         e_to_neighbor_mat.MultTranspose(punity_src_neighbor, e_to_neighbor_avg);
+         fe_dst_to_neighbors_mat.SetRow(i, e_to_neighbor_avg);
+      }
+
+      fe_dst_to_neighbors_mat.InvLeftScaling(neighbors_volume);
+      src_neighbors_avg /=neighbors_volume;
+
+      auto reg_mat = std::make_unique<DenseMatrix>();
+      auto _diff_int = std::make_unique<DiffusionIntegrator>();
+      auto _h1_int = std::make_unique<SumIntegrator>(0);
+      switch (reg_type)
+      {
+         case l2:
+            mass.AssembleElementMatrix(fe_dst_e, e_trans, *reg_mat.get());
+            break;
+         case h1:
+            _h1_int->AddIntegrator(&mass);
+            _h1_int->AddIntegrator(_diff_int.get());
+            _h1_int->AssembleElementMatrix(fe_dst_e, e_trans, *reg_mat.get());
+            break;
+         case direct:
+         default:
+            reg_mat.reset(nullptr);
+            break;
+      }
+
+      if (preserve_volumes)
+      {
+         // Average at e
+         real_t e_volume = mesh.GetElementVolume(e_idx);
+
+         src.GetElementDofValues(e_idx, src_e);
+         punity_src.GetElementDofValues(e_idx, punity_src_e);
+         punity_dst.GetElementDofValues(e_idx, punity_dst_e);
+
+         mass.AssembleElementMatrix(fe_src_e, e_trans, e_to_e_mat);
+
+         real_t src_e_avg = e_to_e_mat.InnerProduct(src_e, punity_src_e);
+         src_e_avg /= e_volume;
+
+         // (Average) projector
+         mass.AssembleElementMatrix(fe_dst_e, e_trans, e_to_e_mat);
+
+         e_to_e_mat.MultTranspose(punity_dst_e, e_to_e_avg);
+         e_to_e_avg /= e_volume;
+
+         LSSolver(solver, fe_dst_to_neighbors_mat, e_to_e_avg, src_neighbors_avg,
+                  src_e_avg, dst_e, reg, reg_mat.get());
+      }
+      else
+      {
+         LSSolver(solver, fe_dst_to_neighbors_mat, src_neighbors_avg, dst_e, reg,
+                  reg_mat.get());
+      }
+
+      if (auto iter_solver = dynamic_cast<IterativeSolver*>(&solver))
+      {
+         if (!iter_solver->GetConverged())
+         {
+            mfem_error("\n\tIterative solver failed to converge!");
+         }
+      }
+      if (print_level >= 0) { CheckLSSolver(fe_dst_to_neighbors_mat, src_neighbor, dst_e); }
+
+      dst.SetSubVector(e_dst_dofs, dst_e);
+
+      e_dst_dofs.DeleteAll();
+      neighbors_e.DeleteAll();
+   }
+}
+
+/// @brief An out-of-element, L2-based reconstruction
```

Comment:
Please expand the comments here to describe the optimization problem being solved and the method for solving it (see comments on LSSolver)